### PR TITLE
Fix/supercharged

### PR DESCRIPTION
**Pull request by tristan-ranff**
- remove truncate (and np.truncate)
- vastly enhanced performance of composition mass and offset calculation (now 2-3x faster)

### CODE REVIEW

File: pyprotista/parsers/misc.py [hunk 1]
```diff
@@ -1,113 +1,83 @@
...
-        )
-        isotopologue_acc = (exp_mz - isotopologue_mz) / isotopologue_mz * 1e6
-    except (KeyError, Exception):
-        logger.warning(f"Could not calculate mz for {seq}#{mods}")
-        pass
-    return composition, c12_mass, isotopologue_acc
+                        m
+                        for m in symbol_to_masses[element]
+                        if str(round(m)).startswith(mass)
+                    ][0]
```
Comment (fu) [inline pyprotista/parsers/misc.py:53]: Reads awkward - what does it do? I guess 14N ? I would suggest a different data structure ... Maybe wrong - phone / train / again

Comment (fu) [inline pyprotista/parsers/misc.py:53]: Reads awkward - what does it do? I guess 14N ? I would suggest a different data structure ... Maybe wrong - phone / train / again

**Review by fu — CHANGES_REQUESTED** [review]

File: pyprotista/parsers/ident_base_parser.py [hunk 2]
```diff
@@ -248,31 +258,117 @@ def check_enzyme_specificity(self):
...
+            ordered_element_multiplier = np.array(ordered_element_multiplier)
+            atom_counts += np.outer(
+                np.char.count(sequences, aa), ordered_element_multiplier
+            )
+        # Remove water (peptide bonds)
+        water = np.zeros(shape=(1, len(elements)), dtype=int)
+        water[0, elements.index("H")] = 2
+        water[0, elements.index("O")] = 1
+        atom_counts -= np.outer(np.maximum(np.char.str_len(sequences) - 1, 0), water)
+        for mod in self.mod_dict.keys():
```
[the remark this answers is not in the document]
Comment (tristan-ranff) [reply pyprotista/parsers/ident_base_parser.py:301]: only copied logic code in cc H2O is removed per added amino acid individually

File: pyprotista/parsers/ident_base_parser.py [hunk 3]
```diff
@@ -281,24 +377,39 @@ def calc_masses_offsets_and_composition(self):
...
-                rt_lookup["rt"].apply(trunc, args=(self.rt_truncate_precision,)),
-            ],
-            inplace=True,
-        )
+        rt_lookup = {}
+        with open(self.params["rt_pickle_name"], mode="r") as meta_csv:
+            meta_reader = csv.DictReader(meta_csv)
+            for row in meta_reader:
+                rt = float(row["rt"])
+                if row["rt_unit"] == "minute" or row["rt_unit"] == "min":
```
[the remark this answers is not in the document]
Comment (tristan-ranff) [reply pyprotista/parsers/ident_base_parser.py:406]: if there is a column specifying the unit for each row, then I am expecting changes

File: pyprotista/parsers/misc.py (hunk 1 above)
Comment (tristan-ranff) [reply pyprotista/parsers/misc.py:53]: should be addressed in future pr in conjunction with the now already present atom count

**Review by MKoesters — APPROVED** [review]

**Pull request merged**